### ISSUE TITLE
Initial support for Mi Band 2

### ISSUE THREAD

**Issue by cpfeiffer** [Freeyourgadget/Gadgetbridge#323]

**Comment by cpfeiffer** [comment]
Anyone having an actual Mi Band 2 is welcome to test and report. If it is compatible to the previous bands, it should just work as is (except for firmware updates). (Gadgetbridge version 0.10.0 will display '?' as device; newer versions will properly recognize the band.

**Comment by santalaus** [comment]
Hello,
got my Band yesterday and tryed to get it running. Bluetooth connect did not work. The Band 2 were not found weather it was connected with Mi Fit or not.
Log Files only have the BT Connection tryes. How could I help.
Best regards.
Nico

**Comment by cpfeiffer** [comment]
Hi Nico, thanks for your testing! :+1: 

Did the scanning dialog in Gadgetbridge show the device at all? If it wasn't shown, they probably have a different Mac address. At the moment we look for devices with Mac addresses that start either with
`88:0F:10:` or `C8:0F:10:` Is yours among them? If it starts with different characters, could you post them here?

A log file captured with Gadgetbridge might still help (see https://github.com/Freeyourgadget/Gadgetbridge/wiki/Log-Files)

Otherwise two separate bluetooth snoops, one created with MiFit connecting, and the other created with Gadgetbrige connecting might help (use the developer settings in Android's settings pane to enable bluetooth dumps. If you don't want to share them publically, I can provide a private upload location.

**Comment by ashimokawa** [comment]
> Ok, my mi band 2 mac address is starting from C2:84:B2:D5

Source: https://www.reddit.com/r/miband/comments/4njl59/just_got_my_mi_band_2_ask_me_anything_d/

**Comment by santalaus** [comment]
Hallo
Mac started with FC:B3:B5:
Band were not shown/found 
Best regards

Am 19. Juni 2016 23:53:37 MESZ, schrieb Carsten Pfeiffer <notifications@github.com>:
>Hi Nico, thanks for your testing! :+1: 
>
>Did the scanning dialog in Gadgetbridge show the device at all? If it
>wasn't shown, they probably have a different Mac address. At the moment
>we look for devices with Mac addresses that start either with
>`88:0F:10:` or `C8:0F:10:` Is yours among them? If it starts with
>different characters, could you post them here?
>
>A log file captured with Gadgetbridge might still help (see
>https://github.com/Freeyourgadget/Gadgetbridge/wiki/Log-Files)
>
>Otherwise two separate bluetooth snoops, one created with MiFit
>connecting, and the other created with Gadgetbrige connecting might
>help (use the developer settings in Android's settings pane to enable
>bluetooth dumps. If you don't want to share them publically, I can
>provide a private upload location.
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Freeyourgadget/Gadgetbridge/issues/323#issuecomment-227022602

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by danielegobbetti** [comment]
I'm fairly sure that the device reports itself with a random MAC Address to unpaired (unbonded) BT peers. It's a standard feature nowadays: [Source](https://e2e.ti.com/support/wireless_connectivity/bluetooth_low_energy/f/538/t/247015).

If we still want to filter, we need to do it on device name or some other feature. Can you please install BLE Explorer from Fdroid and tell us what you see when using that app?

**Comment by Bahjesaispas** [comment]
For me BLE Explorer have took à really long time to detect my Mi band 2 
http://hpics.li/aace680
http://hpics.li/c6060b3

For more pics dont hesitate

**Comment by cpfeiffer** [comment]
Please try the next release. If nothing else, it should discover your mi 2 and allow you to connect it.
If connection still does not work, I'd love to see some log output.

**Comment by hovancik** [comment]
Hi, just got my mibadn 2. Latest version from fdroid: 

```
08:47:07.128 [main] INFO  n.f.g.Logging - Gadgetbridge version: 0.11.1
08:47:07.251 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:07.295 [main] DEBUG n.f.g.s.DeviceCommunicationService - DeviceCommunicationService is being created
08:47:07.295 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.start
08:47:07.304 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.request_deviceinfo
08:47:10.220 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.connect
08:47:10.227 [main] INFO  n.f.g.s.b.BtLEQueue - Attempting to connect to MI Band 2
08:47:10.228 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - Queue Dispatch Thread started.
08:47:10.239 [main] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTING
08:47:10.265 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:10.274 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:47:10.795 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - connection state change, newState: 2 (success)
08:47:10.796 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Connected to GATT server.
08:47:10.797 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTED
08:47:10.800 [main] INFO  n.f.g.e.BluetoothConnectReceiver - got connection attempt
08:47:10.814 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Attempting to start service discovery:true
08:47:10.817 [main] INFO  n.f.g.e.BluetoothConnectReceiver - won't connect to DA:55:B3:07:50:6F(MI Band 2)
08:47:10.834 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:10.839 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:47:10.840 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:47:10.840 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to write characteristic: null
08:47:10.841 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:47:10.841 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to pair MI device...
08:47:10.841 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Unable to pair MI device -- characteristic not available
08:47:10.841 [Binder_1] DEBUG n.f.g.s.d.m.MiBandSupport - Requesting Device Info!
08:47:10.842 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:47:10.862 [Binder_1] DEBUG n.f.g.s.d.m.MiBandSupport - Writing User Info!
08:47:10.862 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to set wear location...
08:47:10.863 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set Wear Location
08:47:10.863 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to set Fitness Goal...
08:47:10.864 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set Fitness Goal
08:47:10.864 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:47:10.865 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:47:10.865 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:47:10.865 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:47:10.874 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Sending current time to Mi Band: July 12 (12 Jul 2016 06:47:10 GMT)
08:47:10.875 [Binder_1] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set time -- characteristic not available
08:47:10.875 [Binder_1] DEBUG n.f.g.s.d.m.MiBandSupport - Requesting Battery Info!
08:47:10.875 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:47:10.876 [Binder_1] WARN  n.f.g.s.b.TransactionBuilder - Unable to write characteristic: null
08:47:10.876 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - about to add: 8:47:10 AM: Transaction task: Initializing device with 7 actions
08:47:10.885 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:47 AM: SetDeviceStateAction to INITIALIZING
08:47:10.889 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:47 AM: ReadAction on characteristic: 00002a00-0000-1000-8000-00805f9b34fb
08:47:10.893 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:10.896 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:47:11.179 [Binder_3] DEBUG n.f.g.s.b.BtLEQueue - characteristic read: 00002a00-0000-1000-8000-00805f9b34fb (success)
08:47:11.180 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:47 AM:  on characteristic: (null)
08:47:11.182 [GadgetBridge GATT Dispatcher] ERROR n.f.g.s.b.BtLEQueue - Queue Dispatch Thread died: Attempt to invoke virtual method 'boolean nodomain.freeyourgadget.gadgetbridge.service.devices.miband.DeviceInfo.isMili1()' on a null object reference
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean nodomain.freeyourgadget.gadgetbridge.service.devices.miband.DeviceInfo.isMili1()' on a null object reference
	at nodomain.freeyourgadget.gadgetbridge.devices.miband.UserInfo.getData(UserInfo.java:88) ~[na:0.0]
	at nodomain.freeyourgadget.gadgetbridge.service.devices.miband.MiBandSupport$1.run(MiBandSupport.java:267) ~[na:0.0]
	at nodomain.freeyourgadget.gadgetbridge.service.btle.BtLEQueue$1.run(BtLEQueue.java:88) ~[na:0.0]
08:47:11.182 [GadgetBridge GATT Dispatcher] INFO  n.f.g.s.b.BtLEQueue - Queue Dispatch Thread terminated.
08:47:41.265 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - connection state change, newState: 0 (failed: 19)
08:47:41.265 [Binder_1] WARN  n.f.g.s.b.BtLEQueue - connection state event with error status 19
08:47:41.265 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Disconnected from GATT server.
08:47:41.266 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - handleDisconnected: 19
08:47:41.266 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - internal gatt callback set to null
08:47:41.266 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: NOT_CONNECTED
08:47:41.272 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - disconnect()
08:47:41.275 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Disconnecting BtLEQueue from GATT device
08:47:41.295 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:41.295 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: NOT_CONNECTED
08:47:41.303 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:47:41.313 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:47:41.317 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:03.451 [main] INFO  n.f.g.e.AlarmReceiver - won't send sunrise and sunset events (disabled in preferences)
08:48:10.882 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.connect
08:48:10.889 [main] DEBUG n.f.g.s.b.BtLEQueue - disconnect()
08:48:10.893 [main] INFO  n.f.g.s.b.BtLEQueue - Attempting to connect to MI Band 2
08:48:10.894 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - Queue Dispatch Thread started.
08:48:10.904 [main] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTING
08:48:10.918 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:10.923 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:14.072 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - connection state change, newState: 2 (success)
08:48:14.073 [Binder_2] INFO  n.f.g.s.b.BtLEQueue - Connected to GATT server.
08:48:14.074 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTED
08:48:14.079 [main] INFO  n.f.g.e.BluetoothConnectReceiver - got connection attempt
08:48:14.086 [Binder_2] INFO  n.f.g.s.b.BtLEQueue - Attempting to start service discovery:true
08:48:14.100 [main] INFO  n.f.g.e.BluetoothConnectReceiver - won't connect to DA:55:B3:07:50:6F(MI Band 2)
08:48:14.109 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:14.109 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to write characteristic: null
08:48:14.109 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:48:14.112 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to pair MI device...
08:48:14.112 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Unable to pair MI device -- characteristic not available
08:48:14.112 [Binder_2] DEBUG n.f.g.s.d.m.MiBandSupport - Requesting Device Info!
08:48:14.112 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:48:14.113 [Binder_2] DEBUG n.f.g.s.d.m.MiBandSupport - Writing User Info!
08:48:14.113 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to set wear location...
08:48:14.113 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set Wear Location
08:48:14.113 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to set Fitness Goal...
08:48:14.113 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set Fitness Goal
08:48:14.114 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:14.114 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:14.114 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:14.114 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:14.115 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Sending current time to Mi Band: July 12 (12 Jul 2016 06:48:14 GMT)
08:48:14.116 [Binder_2] INFO  n.f.g.s.d.m.MiBandSupport - Unable to set time -- characteristic not available
08:48:14.118 [Binder_2] DEBUG n.f.g.s.d.m.MiBandSupport - Requesting Battery Info!
08:48:14.118 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:48:14.118 [Binder_2] WARN  n.f.g.s.b.TransactionBuilder - Unable to write characteristic: null
08:48:14.119 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - about to add: 8:48:14 AM: Transaction task: Initializing device with 7 actions
08:48:14.120 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:48 AM: SetDeviceStateAction to INITIALIZING
08:48:14.121 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:48 AM: ReadAction on characteristic: 00002a00-0000-1000-8000-00805f9b34fb
08:48:14.121 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:14.125 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:14.155 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:14.160 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:14.456 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - characteristic read: 00002a00-0000-1000-8000-00805f9b34fb (success)
08:48:14.458 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - About to run action: July 12, 8:48 AM:  on characteristic: (null)
08:48:14.459 [GadgetBridge GATT Dispatcher] ERROR n.f.g.s.b.BtLEQueue - Queue Dispatch Thread died: Attempt to invoke virtual method 'boolean nodomain.freeyourgadget.gadgetbridge.service.devices.miband.DeviceInfo.isMili1()' on a null object reference
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean nodomain.freeyourgadget.gadgetbridge.service.devices.miband.DeviceInfo.isMili1()' on a null object reference
	at nodomain.freeyourgadget.gadgetbridge.devices.miband.UserInfo.getData(UserInfo.java:88) ~[na:0.0]
	at nodomain.freeyourgadget.gadgetbridge.service.devices.miband.MiBandSupport$1.run(MiBandSupport.java:267) ~[na:0.0]
	at nodomain.freeyourgadget.gadgetbridge.service.btle.BtLEQueue$1.run(BtLEQueue.java:88) ~[na:0.0]
08:48:14.460 [GadgetBridge GATT Dispatcher] INFO  n.f.g.s.b.BtLEQueue - Queue Dispatch Thread terminated.
08:48:44.497 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - connection state change, newState: 0 (failed: 19)
08:48:44.500 [Binder_2] WARN  n.f.g.s.b.BtLEQueue - connection state event with error status 19
08:48:44.501 [Binder_2] INFO  n.f.g.s.b.BtLEQueue - Disconnected from GATT server.
08:48:44.523 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - handleDisconnected: 19
08:48:44.524 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - internal gatt callback set to null
08:48:44.541 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: NOT_CONNECTED
08:48:44.542 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - disconnect()
08:48:44.544 [Binder_2] INFO  n.f.g.s.b.BtLEQueue - Disconnecting BtLEQueue from GATT device
08:48:44.552 [Binder_2] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: NOT_CONNECTED
08:48:44.572 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:44.575 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:44.593 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:44.599 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:47.309 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:47.333 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.start
08:48:47.335 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.request_deviceinfo
08:48:47.373 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:47.392 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:48.854 [main] DEBUG n.f.g.s.DeviceCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbridge.devices.action.connect
08:48:48.855 [main] DEBUG n.f.g.s.b.BtLEQueue - disconnect()
08:48:48.858 [main] INFO  n.f.g.s.b.BtLEQueue - Attempting to connect to MI Band 2
08:48:48.861 [main] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTING
08:48:48.861 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:48.862 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.s.b.BtLEQueue - Queue Dispatch Thread started.
08:48:48.884 [main] WARN  n.f.g.i.GBDeviceCandidate - no cached services available for MI Band 2: DA:55:B3:07:50:6F
08:48:49.169 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - connection state change, newState: 2 (success)
08:48:49.171 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Connected to GATT server.
08:48:49.172 [Binder_1] DEBUG n.f.g.s.b.BtLEQueue - new device connection state: CONNECTED
08:48:49.172 [main] INFO  n.f.g.e.BluetoothConnectReceiver - got connection attempt
08:48:49.175 [Binder_1] INFO  n.f.g.s.b.BtLEQueue - Attempting to start service discovery:true
08:48:49.180 [Binder_3] WARN  n.f.g.s.b.TransactionBuilder - Unable to notify characteristic: null
08:48:49.181 [Binder_3] WARN  n.f.g.s.b.TransactionBuilder - Unable to write characteristic: null
08:48:49.181 [Binder_3] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:48:49.181 [Binder_3] INFO  n.f.g.s.d.m.MiBandSupport - Attempting to pair MI device...
08:48:49.180 [main] INFO  n.f.g.e.BluetoothConnectReceiver - won't connect to DA:55:B3:07:50:6F(MI Band 2)
08:48:49.182 [Binder_3] INFO  n.f.g.s.d.m.MiBandSupport - Unable to pair MI device -- characteristic not available
08:48:49.182 [main] INFO  n.f.g.s.DeviceCommunicationService - Setting broadcast receivers to: true
08:48:49.182 [Binder_3] DEBUG n.f.g.s.d.m.MiBandSupport - Requesting Device Info!
08:48:49.183 [Binder_3] WARN  n.f.g.s.b.TransactionBuilder - Unable to read characteristic: null
08:48:49.183 [Binder_3] DEBUG n.f.g.s.d.m.MiBandSupport - Writing User Info!
```

**Comment by cpfeiffer** [comment]
Now that #349 is fixed, I'd like to get Mi Band 2 working. Without having a device, I could need some help :-) If anyone of you can build Gadgetbridge yourself so we can try things out without the whole release -> f-droid -> install cycle, we will probably get something done, quickly.

At the moment we fail because we're expecting to get some services/characteristics that are not available.

Very helpful would be a 
- a list of all its ble services and characteristics (the screenshots from @Bahjesaispas only show parts of it)
- Mi Fit logfile
- a Mi Fit btsnoop log of the connection process

**Comment by sir-harlekin** [comment]
i will provide BLE screenshots tomorrow.
where can i find the mi fit logfile? (will look for myself tomorrow, maybe an answer is not needed)

**Comment by cpfeiffer** [comment]
It could be available as /sdcard/mili_log.txt. Not sure if it is created in all cases or if it has to be enabled in the app.

**Comment by sir-harlekin** [comment]
can i send you the logfile somehow? it now is 4mb (up from the first start of the app)

**Comment by sir-harlekin** [comment]
Log file (4 mb, up from the first start of the app) attached!
[mili_log_20160715222608.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/366801/mili_log_20160715222608.zip)

**Comment by sir-harlekin** [comment]
All screenshots i could take from BLE explorer (may contain slight
[BLE.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/366828/BLE.zip)
 duplicates)

**Comment by cpfeiffer** [comment]
Awesome, thanks!

**Comment by cpfeiffer** [comment]
From a quick look, it seems as they changed a lot. The current Mi Band 1 support will not work with it. Maybe we can reuse some code of it, but it basically looks like a completely new device that needs to be understood.

**Comment by sir-harlekin** [comment]
Can i provide something else to support? I have slide problems to understand, what sniffing BT data means or how to accomplish that with the knowledge of an 'average' user.

**Comment by cpfeiffer** [comment]
I will document how to capture a btsnoop log in the wiki, but for short, go to Android Settings -> Developer Options -> Enable Btsnoop (Bluetooth Capture). Then perform a connect with MiFit and disable Btsnoop (so that the only contains the connection information). You will get a btsnoop_hci.log in your /sdcard/ directory.

Thanks!

**Comment by sir-harlekin** [comment]
So here is what i got. I am not sure, if i have done this the right way, because it looks a lot like gibberish to me. Please check and come back if i should try again.

[btsnoop_hci.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/367188/btsnoop_hci.zip)

**Comment by santalaus** [comment]
Hey Carsten,
would it help if you get one Device? I have one left here and you were also an German Guy so sending would be easy. Maybe the other Guys were interested in getting support and we could have something line croudfoundig an device for you. To make it quick I would send it directly. How were Interestes to send me some Bucks could send me an private message. The device cost me right now 45€ I think we need 5€ to send it to you.
Best regards.
Nico

**Comment by cpfeiffer** [comment]
@sir-harlekin Yes, it looks like gibberish, but that's fine :-)
You (I) have to decode it with
```hcidump -axt -r btsnoop_hci.log```
Then it still looks like gibberish, but a lot less :-)

Thanks a lot, I will try to make some sense out of it.

**Comment by cpfeiffer** [comment]
@santalaus Yes, that would certainly speed things up :-) In case you're near or in Berlin, we could spare the shipping costs and I could also return it quickly in case you need it again.

**Comment by santalaus** [comment]
@cpfeiffer sorry, I live near Dortmund and in the next time my way did not came to Berlin. So shiüüing could be useful. Offer is there.
Best Regards Nico

**Comment by sir-harlekin** [comment]
If i can help in any way give me a beep. Btw: I live exactly in the middle between B and DO :-)

**Comment by cpfeiffer** [comment]
@santalaus I can give you my address, and I can send it back to you as soon as you need it. If it breaks, I wouldn't order another, though. I have 6 Mi Bands (1, 1A, 1S) and so far,
- one has a somewhat bad vibration motor (since when I received it)
- one does not record any activity anymore (live activity still works) and firmware upgrade does not work anymore
- one sporadically does not record activity, but most of the time does
- one does record activity data, but only a few hours of data are good, the rest is bogus

Exactly two out of 6 are still working as they should.  I kept all of them dry, they didn't fall down or anything. So it seems to me that the quality of the memory of Mi Band 1 models is really bad.

Maybe the Mi Band 2 is better quality-wise, but I cannot guarantee anything.

**Comment by santalaus** [comment]
Hello,
I have got 2 Mi Band 2 one used with original SW and one unused. So no problem ;)
The offer sendeing it to you is my intention to get support for it, you could have it.

Maybe the other guys want to share the costs with me, but if I get your address (I have now some problems with internet access), I would send it to you so you could have it as Thank you for developing..

**Comment by cpfeiffer** [comment]
Alright, send me a mail to carsten@freeyourgadget.org, then I will give you my postal address. Or if you have a jabber ID, we could use that as well. Looking forward to Mi 2 support :-)

Note, I don't mind if you send the used or unused one.

**Comment by santalaus** [comment]
You should have mail.

**Comment by cpfeiffer** [comment]
Got it, thanks :-)

**Comment by sir-harlekin** [comment]
Can we further assists on Mi Band 2 compatibility?

**Comment by santalaus** [comment]
Post Delivery were successful, Now give him some time. He wrote me, that the Band 2 were completly different from 1

**Comment by martin-v** [comment]
Hi @cpfeiffer, can I help you with this issue? Tomorrow will probably my new Mi Band 2 arrive, then can I help you to test or write some code ...

**Comment by cpfeiffer** [comment]
@martin-v sure, all help is appreciated. Our current state and findings are:
- all activity happens in the master branch
- Mi Band 2 is a very different beast than Mi Band 1x, it is actually more standards compliant AFAICS
- due to that, we started to implement the standard ble services/profiles in a device independent way
- for a start, we have device information and battery
- **we need support for the current the time service** in order to send the current time to the band
- see https://www.bluetooth.com/specifications/adopted-specifications (CTS)
- the important classes for devices are the DeviceSupport classes, in this case `MiBandSupport` (for Mi Band1) and `MiBand2Support`. This is just for testing, so for now, you will need to change the single constructor invocation of `MiBandSupport` to instantiate `MiBand2Support` instead
- then you will be able to connect and get some device information (hw revision) from it
- the device will close the connection after a few seconds because we do not act as expected
- we need to find out, what to send to the band after getting device information. current time for sure, but from our experience with Mi Band 1, we will probably need to send some user information to the band (alias, height, weight, age, ...), see class UserInfo, but we don't know how it's done with Mi Band 2

If you know how to read Mi Fit logs/bluetooth snoops, e.g. with hcidump or wireshark, maybe you can find out what to send to the band. Note, the the GB logfile (or Android Monitor in Android Studio) will have helpful information, like the discovered, services, characteristics, etc.

Let me know if I can assist anywhere.

**Comment by martin-v** [comment]
Thanks for the information's, I have the same status now, I can connect, but the connection is closed after a few seconds. I will try implement the current the time service, but I think i need some time to understand the Code base and the Bluetooth Specs.

**Comment by diimdeep** [comment]
Hello I have mi band2 might try to help with it, and I have another device bong3hr.

**Comment by cpfeiffer** [comment]
Sorry for the late response -- we had to temporarily close the issue tracker because we were overloaded with requests and discussions and thus couldn't complete the big db changes.

Thanks to @JohnnySun, we now have a much better undestanding of Mi Band 2 and even some working code. We have a stable connection, and we can change the time of the band. Heart rate readings and notifications are probably easy. Firmware updates and activity data reading is probably harder.

All help and support is appreciated :-)

**Comment by cpfeiffer** [comment]
So next steps are implementing more standard ble services (heart rate profile, alert notification profile).

**Comment by Almtesh** [comment]
Hi,
I installed the 0.12.2 version from F-droid and the connection does not succed. Is it in the next version?

**Comment by cpfeiffer** [comment]
Yes, it will be on the next release.

**Comment by Almtesh** [comment]
If you need testers, I can help you. Badly, I can't understand Java, so I can't help for dev, but I can test.

**Comment by ashimokawa** [comment]
We have just tagged 0.13.0, which will be on f-droid soon and have the initial Mi Band 2 support (no activity, no heart rate, just notifications for now)

**Comment by Almtesh** [comment]
These missing feature are few price for liberty, good luck for implementing the rest.

**Comment by ashimokawa** [comment]
0.13.0 is out in f-droid now.

**Comment by Almtesh** [comment]
Yes, I updated it. There is only one type of notification for the moment?

**Comment by cpfeiffer** [comment]
They use the same icon right now, but different vibration patterns.

**Comment by cafilho** [comment]
Hi. My Mi Band 2 stopped working because I deleted my Mi Account without parring. Now it shows the "Band is paired by another person. Ask them to unpair it and try again" message.
I found the Gadgetbridge project when trying another method to use my band again (still no success).

...long story short...

I've bought a new Mi Band 2 and it will arrive at any time. I know that when I first pair to Mi Fit it will update it's firmware. 
If it is of any help, I can Log Bluetooth, network, processes, etc and share the logs here. I just need some one to teach me how to do it. My phone is running Android 6.0 without root.

**Comment by cpfeiffer** [comment]
@Almtesh of you mean the icon on the band, then yes. You can configure different vibration patterns, though.

**Comment by cpfeiffer** [comment]
@cafilho So Gadgetbridge was unable to take over your band?

Before throwing the band away, you could try completely discharging the battery, e.g. by putting it into the freezer for a few days. Not sure if the display like that, though, so maybe the fridge or just waiting for the discharge would be safer.

Yes, logs do help a lot. You can enable a BT snoop in android's development settings. Everything you do with Bluetooth afterwards will then be logged to a file btsnoop.log in /sdcard.

**Comment by santalaus** [comment]
Hello,
I'm on the latest Version from F-Droid, but sometimes (once a day, i think) the Band were disconnected. I have to reconnect it manual to get Messages again.
Did anyone also have this, or is it a Problem of my configuration?
Best regards
Nico

**Comment by Almtesh** [comment]
Is the autoconnect setting set? Pearhaps your app is not set to autoconnect when disconnect happen.

**Comment by santalaus** [comment]
Autoconnect is on. Also connect when Bluetooth came up.

**Comment by cpfeiffer** [comment]
@santalaus Maybe the device initialization gets stuck somewhere... if you could provide a log when that happens, maybe we could find out what's happening.

**Comment by cpfeiffer** [comment]
@JohnnySun Do you have some more information regarding the used services/characteristics? E.g for setting alarms, fetching activity data, etc.

**Comment by santalaus** [comment]
@cpfeiffer BT-Log or Application Log? From my feeling this happens after an alarm. But till now I only see when this happend time ago. I could not reproduce it. I just klick ob MiBand2 and it reconects.

Have you seen the new feature in the MiBand Firmware? Changing Time Display to Time and Date.

**Comment by cpfeiffer** [comment]
I meant an application log.

**Comment by cpfeiffer** [comment]
No, I didn't see the new firmware yet, I'm not running Mi Fit, even. Fell free to share some BT logs of specific things that we should implement. We lack hr reading, activity reading, setting alarms and fw update, for example.

**Comment by santalaus** [comment]
FYI done an Update and aktivated the Logfile.
MI Band FW: V1.0.0.39 HW: V0.1.3.2
Now not using Mi Fit anymore. So for BT logs I have to go back and have to read/learn how to do BT logs.

**Comment by cpfeiffer** [comment]
I added some info how to capture Bluetooth traffic to https://github.com/Freeyourgadget/Gadgetbridge/wiki/Log-Files

**Comment by 6arms1leg** [comment]
I took the liberty to modify/add information to above wiki page – I hope that's OK.
I also got a Mi Band 2 now and I'm happy to see that support for it is already under development.
Sadly, my new Mi Band 2 almost never connects, even the initial pairing process almost always fails.
Strangely, after a failed pairing attempt my Mi Band 2 still shows on the main screen but I can't connect (the icon becomes blue, the state switches to "connecting" but it never finished/switch to "connected").

Here is a BT snoop log of the failing pairing process with Gadgetbridge:
[btsnoop_hci-init_fail.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/518324/btsnoop_hci-init_fail.log.zip)

Here is a BT snoop log of the failing connecting process with Gadgetbridge:
[btsnoop_hci-connect_fail.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/518332/btsnoop_hci-connect_fail.log.zip)

The pairing process was successful only once (I have no idea why). After that, I could connect the paired device.

Here is a BT snoop log of the successful pairing process with Gadgetbridge:
[btsnoop_hci-init.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/518333/btsnoop_hci-init.log.zip)

Here is a BT snoop log of the successful connecting process with Gadgetbridge:
[btsnoop_hci-connect.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/518334/btsnoop_hci-connect.log.zip)

I deleted the Mi Band 2 from Gadgetbridge after that again and since then I wasn't able to bond the Mi Band 2 again. I also tried on a different device with Xiaomi's Mi Fit app to provide some logs, but the app doesn't even find the Mi Band 2. Could that be because Gadgetbridge is not correctly unpairing the Mi Band 2 and so Xiaomi's Mi Fit app doesn't show the device at all (because it's still bonded with Gadgetbridge on my other device)?
I couldn't generate a Gadgedbridge log file due to #406.
I hope some of this might be helpful.

**Comment by cpfeiffer** [comment]
Thanks for updating the wiki!

I only have a single Mi Band 2, and connecting works perfectly with my One Plus One. I'll test connection with another device. A gadgetbridge log would help a lot -- I hope we will find a way to make that work for you.

Note that discovering the device may take a little while, sometimes even multiple attempts.

**Comment by 6arms1leg** [comment]
Discovering the device itself is never a problem – it always shows up in the list on the discovery screen as "Mi Band 2". But then selecting it to bond with it almost always fails (in fact, it only worked once).

Since Gadgetbridge's log feature is currently not working on my device, I have logged the following actions with logcat (see attached file at the bottom):

- Trying to add the Mi Band 2 (by tapping the blue "+" button on the lower right corner in the control center/main screen and selecting the new device from the list on the discovery screen)
- Aborting the action by going back to the control center/main screen (since step one never finishes)
- Oddly, the new device still shows up in the control center/main screen, so I try to connect the device
- deleting the device again (after connecting fails)

logcat log file:
[catlog-failed_pairing_failed_connecting_delete_dev.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/518577/catlog-failed_pairing_failed_connecting_delete_dev.log.zip)

**Comment by cpfeiffer** [comment]
Interesting. It looks like the code for Mi Band 1 is used instead of the code for Mi Band 2. So no wonder it doesn't work. We need to fix something in the device detection, I suppose.

**Comment by cpfeiffer** [comment]
I have an idea. Please force-close Gadgetbridge and try connecting again. Make sure to *not* connect to your Mi Band 1 before!

**Comment by 6arms1leg** [comment]
Great idea! This seems to fix it. I tried several times and each time it worked perfectly that way. Thanks!

**Comment by santalaus** [comment]
Short response: Since the last update the connection seemed to be stable. So no App-Log from my side.
Second: I would try to find some time in the next Days to take the Bluetooth Logs. Could I make one big Log or is it better one action per log (If I could)?

**Comment by cpfeiffer** [comment]
Great! Yes, definitely one log per action so that we can correlate the data to the expected actions.

**Comment by santalaus** [comment]
[btsnoop_hci.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/529481/btsnoop_hci.zip)
Here some Work for your Weekend ;)
Timeformat -Display just Clock, Clock+Date
Displaysettings alloff allon Time(allwayson),Steps,Distance,Kalorien,Hartrate,Remaining Battery
HeartRateSleepAssistent off/on

**Comment by 6arms1leg** [comment]
Did you already pushed the latest release to F-Droid? Somehow I can't update nor find the newest version on the F-Droid website... Many thanks!

**Comment by cpfeiffer** [comment]
I pushed the tag 4 days ago. That's indeed quite a long time.

**Comment by ashimokawa** [comment]
Yes we did days ago.
F-droid wiki says they have the latest version. But they don't
https://f-droid.org/wiki/page/nodomain.freeyourgadget.gadgetbridge

**Comment by santalaus** [comment]
Sorry for the confussion. The latest from F-Droid

[btsnoop_hci2.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/531164/btsnoop_hci2.zip)

Some more Bluetooth logs.
- Display of Alarm Symbols everyone once triggert
- donotdisburt set off/on of Donotdisburt Time (Dont know if there is Bluetooth Traffic)
- MessureHeartrate Messuring of Heartrate
- NotifyCall set on/off (non call triggert)

Hope this would help
My wish for an funktion would be: Alarm sleeping times and an Alarm Whitelist instead of Blacklist.
Also possibility of selection Icon / vibration per App.

Best regards

**Comment by cafilho** [comment]
Hi!

As  I promised, my new mi band arived and here is the log of first connection with mi fit and firmware update (and, probably, date/hour set).

I can produce more logs if requested.
[btsnoop_hci(first conection and firmware update).zip](https://github.com/Freeyourgadget/Gadgetbridge/files/531243/btsnoop_hci.first.conection.and.firmware.update.zip)

**Comment by cpfeiffer** [comment]
Thanks a lot, guys. I didn't have time to check them this weekend, but will do so in the next days!

**Comment by FichteFoll** [comment]
My Mi Band 2 arrived a few days ago and I was eager to try it out with GB but unfortunately I'm unable to pair. The device just never shows up and eventually my phone restarts after I click on the "+" to add a device. Do I need to do something with the device in order to pair it? I don't find any information about doing anything during the pairing process.

Later I installed the Mi Fit app but haven't created an account yet.

Not exactly sure how to proceed now, though. I could probably create some logs for the failed pairing attempts, if that helps? Otherwise I'll try to pair with the Mi Fit app (while logging).

**Comment by cpfeiffer** [comment]
Make sure that you have GPS enabled while scanning. This is unfortunately required on Marshmallow. You can safely disable it again afterwards.

**Comment by ashimokawa** [comment]
Network location is enough

**Comment by FichteFoll** [comment]
It was definitely (and still ie) enabled. Only GPS though, not this cellular and WiFi data thing.

**Comment by 6arms1leg** [comment]
Here are some more logs from the Mi Fit app.

Setting alarm to 10am:
[btsnoop_hci-alarm_10am.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533314/btsnoop_hci-alarm_10am.log.zip)

Measuring heart rate (57BPM):
[btsnoop_hci-hr1_57bpm.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533315/btsnoop_hci-hr1_57bpm.log.zip)

Measuring heart rate again (57BPM):
[btsnoop_hci-hr2_57bpm.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533316/btsnoop_hci-hr2_57bpm.log.zip)

Synchronizing activity data:
[btsnoop_hci-sync1.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533318/btsnoop_hci-sync1.log.zip)

Synchronizing activity data again:
[btsnoop_hci-sync2.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533319/btsnoop_hci-sync2.log.zip)

Synchronizing activity data again:
[btsnoop_hci-sync3.log.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533321/btsnoop_hci-sync3.log.zip)

**Comment by FichteFoll** [comment]
I reported my scanning problem to a different issue: #412.

**Comment by FichteFoll** [comment]
Initial pairing and immediately following firmware update with Mi Fit app:

[btsnoop_hci_pair&update.zip](https://github.com/Freeyourgadget/Gadgetbridge/files/533876/btsnoop_hci_pair.update.zip)

**Comment by cpfeiffer** [comment]
More logs, thanks :-)

**Comment by cpfeiffer** [comment]
Short feedback: we needed to write a tool to make the logs readable/understandable/manageable. I had a quick look to implement support for alarms, but the first test wasn't immediately successful.

But progress is happening!

**Comment by cpfeiffer** [comment]
Is there a "Reset Device" or "Reboot Deviice" function in Mi Fit? Min MI2 does not succeed in measuring heart rate anymore, not through Gadgetbridge, nor when pressing the band's button.

@JohnnySun any idea why this might happen or what to do? I'm still on fw 1.0.0.2.

**Comment by santalaus** [comment]
I could not find anything like this.
Maybe with Fw Update but all off my Settings survifed an FW Update.
Best regards.

**Comment by cafilho** [comment]
The only new thing I've found in the new firmware was the option to "rotate the wrist to switch info on Mi Band" (fw v1.0.0.53)

**Comment by 6arms1leg** [comment]
Thanks @cpfeiffer, setting alarms on the Mi Band 2 works great with the new Gadgetbridge version, however turning them off again doesn't.

Also, in [one of your last commits](https://github.com/Freeyourgadget/Gadgetbridge/commit/17b70a1b82a261ee9ab114c86d024419a4da5be5) you remarked:
> Smart alarms not supported yet.

The smart alarm feature seems to be officially discontinued – the new Mi Fit app doesn't have it anymore (at least I can't find it anymore and others reprted that problem too, see [here](http://en.miui.com/thread-267378-1-1.html) and [here](http://c.mi.com/sg/thread-2290-1-1.html)).
The Mi Band Tools app still has it, and it's reported to work with some Mi Band 2s but on the device I tested it, it didn't work. So maybe Xiaomi also removed the feature from newer firmware...

**Comment by 6arms1leg** [comment]
Oh, and about measuring the heart rate:
Measuring the heart rate on the device itself works (most of the time).
In Gadgetbridge, I can now initiate the measurement (the green light from the Mi Band 2's sensor starts to flicker and I get a toast notification about starting the process) but never get a toast about the measured value afterwards.

**Comment by cpfeiffer** [comment]
@6arms1leg thanks, will have a look at disabling alarms. If smart alarms are removed from the firmware, we can't support them, indeed. At the moment, I simply don't know what to send to the band in order to enable them (assuming they are still present in the firmware).

The hr measurements are not released yet, do you build yourself? You need de46555e370cd86b90541e9b7406503e9d66a886 for the toast. If that doesn't work, could you share a gadgetbridge log?

**Comment by 6arms1leg** [comment]
@cpfeiffer, no I use the F-Droid release version. Sorry, my bad! I thought the heart rate feature was already included in the latest release...

**Comment by 6arms1leg** [comment]
Short feedback, @cpfeiffer:
Disabling alarms and measuring heart rate (from the debug screen) works great now on the latest version of GB. As always, thanks!

**Comment by Almtesh** [comment]
I saw I could get the heart rate in debug mode.
Would that be possible to get this from another app (i.e. Termux). I'd like to monitor this with a script.
Thanks.

**Comment by cpfeiffer** [comment]
This is probably not really possible at the moment. If you're root, you might turn on logging in Gadgetbridge and parse the logfile, but I wouldn't suggest doing that ;-)

Would you like to open a new issue for this?

**Comment by Almtesh** [comment]
I openned #435 for that, I allready openned a issue to ask that, but it desapeared.

**Comment by cpfeiffer** [comment]
Thanks. Weird about the disappeared issue :upside_down_face:

**Comment by cpfeiffer** [comment]
Just to report some progress: activity fetching is almost done. It's not pushed yet though, so it be not be part of the next release.

**Comment by Almtesh** [comment]
Nice!

**Comment by TSG9876** [comment]
Hi ,

I am not sure whether the heart rate measurement return value had been resolved or not for mi band 2 , I just share some log below , return value for mi band 2 is (0x) 00-64 as compared to mi pulse MI1S which is (0x)06-64 for example for 100bpm in hex code.


(00002902-0000-1000-8000-00805f9b34fb)
V	07:25:28.564	Reading descriptor 00002902-0000-1000-8000-00805f9b34fb
D	07:25:28.575	gatt.readDescriptor(00002902-0000-1000-8000-00805f9b34fb)
I	07:25:28.650	Read Response received from descr. 00002902-0000-1000-8000-00805f9b34fb, value: (0x) 01-00
A	07:25:28.663	"Notifications enabled" received
V	07:25:29.527	Reading characteristic 00002a39-0000-1000-8000-00805f9b34fb
D	07:25:29.541	gatt.readCharacteristic(00002a39-0000-1000-8000-00805f9b34fb)
I	07:25:29.626	Read Response received from 00002a39-0000-1000-8000-00805f9b34fb, value: (0x) 01-00
A	07:25:29.638	"(0x) 01-00" received
**I	07:25:40.653	Notification received from 00002a37-0000-1000-8000-00805f9b34fb, value: (0x) 00-64**
A	07:25:40.687	"Heart Rate Measurement: 100 bpm,
Sensor Contact Not Supported" received
V	07:25:45.810	Reading characteristic 00002a39-0000-1000-8000-00805f9b34fb
D	07:25:45.820	gatt.readCharacteristic(00002a39-0000-1000-8000-00805f9b34fb)

**Comment by cpfeiffer** [comment]
@TSG9876 I'm not quite sure what you mean with this. Do you mean hr measurement does not work with Gadgetbridge? Or do you mean the value is wrong?

**Comment by TSG9876** [comment]
the log post is non other than just a simple display of the Mi Band 2 heart rate notification return value which take the form of hex code  00-XX as compared to the Mi Band pulse MI1S return hex code in the form as 06-XX as I noticed the heart rate feature seemed still in Alpha or Beta stage for Mi Band 2. 

The Hex byte to trigger the heart rate measurement is still 0x15,1,1 for continuous or  0x15,2,1 for interval measurement reading.

**Comment by cpfeiffer** [comment]
Oh, thanks, I will try that. So far we only sort manual readings via the debug screen and "sleep assistant", i.e. automatic measurements whole sleeping.

The values we receive should be correct AFAICS.

**Comment by TSG9876** [comment]
Mi Band 1S or 2 Heart Notification is received back  from UUID service 00002a37-0000-1000-8000-00805f9b34fb,

To trigger heart rate scanning on demand is done via gatt.write characteristic (00002902-0000-1000-8000-00805f9b34fb) with Hex code 0x15,1,1 for continuous or 0x15,2,1 periodic.

**Comment by TSG9876** [comment]
I saw the amendment ,thanks

 private void handleHeartrate(byte[] value) {
-        if (value.length == 2 && value[0] == 6) { // this is for reading MiBand Pulse MI1S and was removed


+        if (value.length == 2 && value[0] == 0) { // this is to cater for MiBand 2 .
             int hrValue = (value[1] & 0xff);
             int hrValue = (value[1] & 0xff);
             if (LOG.isDebugEnabled()) {
             if (LOG.isDebugEnabled()) {
                 LOG.debug("heart rate: " + hrValue);
                 LOG.debug("heart rate: " + hrValue);
             }

 Just a note on the continuous heart rate measurement trigger by protocol 0x15,0x01,1, is there a noticeable decrease delay in the value feedback something  like after almost 10 times  of consistent less than 1second continuous heart beat feedback , delay for almost 5s before resume normally again with continuous feedback of 1 second repeating the same  cyclical pattern through out the measurement  ?

**Comment by cpfeiffer** [comment]
I haven't really done many measurements with Mi Band 2 yet, so I cannot comment on that. I did notice that the band sends measurement results for quite a while even after telling it to stop measuring.

**Comment by Almtesh** [comment]
I tested and it works well. I don't know if step count works, but I have different values on band and phone. I have 109 steps on GB and 966 on band.

**Comment by 6arms1leg** [comment]
@Almtesh: Yes, I also noticed a difference in the step count between GB and the Mi Band 2 itself with the same trend: The Mi Band 2 displays a higher value than GB.

**Comment by cpfeiffer** [comment]
Thanks for the feedback. The steps have been completely in sync for me, so I'm surprised about this. Will have another look.

**Comment by 6arms1leg** [comment]
I now checked the activity data with my Mi Band 2 (FW 1.0.0.19) on GB (v0.14.3) further and also noticed that no sleep data is displayed. The bar graphs show a gap (no bars at all) were the sleep should be. The pie chart indicates 0min of sleep.

**Comment by Almtesh** [comment]
@cpfeiffer Actually, I installed the update yesterday in the middle of afternoon, so the count starts at this moment, so that's normal it was wrong and it's not anymore.
@6arms1leg I don't have the gap like you, I just got heath rate graph like I want to have for all the day, but there's nothing in the pie chart, like you.

**Comment by Almtesh** [comment]
@6arms1leg did you check the option to get more precise sleep data with heart rate sensors? I checked it, that's perhaps why I got something when I sleep.

**Comment by 6arms1leg** [comment]
@Almtesh, thanks for your answer.

> [...] I installed the update yesterday in the middle of afternoon, so the count starts at this moment, so that's normal it was wrong and it's not anymore.

For me, the step count in GB and on the Mi Band 2 itself still differ, even after midnight (with the Mi Band 2 showing much higher values). :crying_cat_face: 

> [...] I just got heath rate graph [...]. [...] did you check the option to get more precise sleep data with heart rate sensors? I checked it, that's perhaps why I got something when I sleep.

Heart rate readings during sleep are also not shown at all, although the option "Use Heartrate Sensor to improve sleep" in the settings is checked. However, HR reading on the "Debug" screen seems to work (although I get a reading of 0bpm most of the time, but thats the same with HR readings on the Mi Band 2 itself – they also fail most of the time).

**Comment by Almtesh** [comment]
@6arms1leg HR is not really efficient on this arm, try one of your five other. You perhaps have to much hairs between the sensor and your skin, since it's a optical sensor.

**Comment by cpfeiffer** [comment]
Sleep detection is not available in 0.14.3. A first steep towards it will be in 0.14.4, due out soon.

Bit this is really work in progress, as I'm just guessing what the numbers might mean.

**Comment by cpfeiffer** [comment]
My hr sensor stopped working after a while as well, it always displayed some kind of "invalid measurement" sign on the band the value displayed in Gadgetbridge was 0.

This was fixed with a fw update.

The Mi1S models had the same issue by the way. Out of the box, hr measurement did not work at all. It only started working after a fw update.

**Comment by cpfeiffer** [comment]
If someone wants to help implementing support for fw update, please yell. I can give some pointers but lack the time to do it myself right now.

**Comment by cpfeiffer** [comment]
I guess it's time to close this issue, since we're pretty much beyond "initial support".

Activity detection and synchronization is work in progress, firmware update is requested at #427. 

Please open separate issues for further things. Thanks for your support! :heart:

**Comment by phoenix31121** [comment]
Hi, i was having problems connecting band 2 with gb on galaxy s5 6.0.1 mm. It would connect in bluetooth settings but wouldnt authenticate ie no vibration on device. 
But i got hold of the older j5 running 5.0 and gb connected to the band without a hitch.
It seems that newer mm firmwares arent compatible with gb i guess? 
I hope this info helps in any way.

**Comment by 6arms1leg** [comment]
Hello @phoenix31121,

does connecting your Mi Band 2 to your S5 work with the Mi Fit app? Actually, I had the same problem with a friends S5 a few months back. But I couldn't even connect using the Mi Fit app (the Mi Band 2 wasn't paired with any other device), so I figured the issue wasn't GB-related...

BTW, although the issues are piling up I recommend opening a new issue for your reports/questions if you think it's GB-related. Posting to an old closed issue, your post will be easily forgotten/overlooked – a separate issue makes it easier to keep track of it.

**Comment by phoenix31121** [comment]
Hello @6arms1leg, 
Yes the band works perfectly with the mi app. I needed gb to downgrade the firmware as the newer fw almost killed the heart rate sensor. 
Im totally new to gadgetbridge and this fourm and i have no idea how to create a new issue.

**Comment by 6arms1leg** [comment]
> Yes the band works perfectly with the mi app.

Then this might actually be GB-related. You should consider opening a separate issue (by clicking the green button labeled "New issue" in the top right corner of this page or any other issue page). Be prepared to attach a GB log file and/or a BT snoop file to your issue. More information on how to do this can be found in the [wiki](https://github.com/Freeyourgadget/Gadgetbridge/wiki/Log-Files) (it's quite easy).

Welcome to Gadgetbridge and GitHub, BTW!

**Comment by phoenix31121** [comment]
@6arms1leg thank you. i was on mobile the whole time so hadnt seen all the options on the page. i'll try to post the issue. actually im on it now.

**Comment by aytunch** [comment]
@TSG9876 Hi,
you said years ago:)

> V 07:25:28.564 Reading descriptor 00002902-0000-1000-8000-00805f9b34fb
> D 07:25:28.575 gatt.readDescriptor(00002902-0000-1000-8000-00805f9b34fb)
> I 07:25:28.650 Read Response received from descr. 00002902-0000-1000-8000-00805f9b34fb, value: (0x) 01-00
> A 07:25:28.663 "Notifications enabled" received....

and,

> Mi Band 1S or 2 Heart rate Notification is received back from UUID service 00002a37-0000-1000-8000-00805f9b34fb,
> To trigger heart rate scanning on demand is done via gatt.write characteristic (00002902-0000-1000-8000-00805f9b34fb) with Hex code 0x15,1,1 for continuous or 0x15,2,1 periodic.

You say we need to write 0x15,1,1 to **2902...** which is the descriptor for HRM data(or control?). But in the code I can't see any descriptor:

```
@Override
    public void onEnableRealtimeHeartRateMeasurement(boolean enable) {
        if (characteristicHRControlPoint == null) {
            return;
        }
        try {
            TransactionBuilder builder = performInitialized("Enable realtime heart rate measurement");
            enableNotifyHeartRateMeasurements(enable, builder);
            if (enable) {
                builder.write(characteristicHRControlPoint, stopHeartMeasurementManual);
                builder.write(characteristicHRControlPoint, startHeartMeasurementContinuous);
            } else {
                builder.write(characteristicHRControlPoint, stopHeartMeasurementContinuous);
            }
            builder.queue(getQueue());
            enableRealtimeSamplesTimer(enable);
        } catch (IOException ex) {
            LOG.error("Unable to enable realtime heart rate measurement", ex);
        }
    }
```

I can collect bits and pieces of information for getting continuous heart rate from MiBand3(which is same with 2 I think) but I can not get it to work. If you still remember these or point to the code, the exact steps to get continuous HRM data? I read on a blog about needing to send a ping every 12-15 seconds to keep on getting data. But i can not find it in the code either. Thanks a lot.
